thread safety: context vars for batching context

fselmo opened the pull request:
### What was wrong?

Closes #3613

### How was it fixed?

- [bugfix](https://github.com/ethereum/web3.py/pull/3705/commits/2e3559741feba1e305c8ab99ce809de35603900a) Use `ContextVar` rather than global flag to toggle batching context. This should better isolate the batching context to its own thread-safe context.
- Add test for concurrent execution of many requests at the same time as a batch request
- [bugfix](https://github.com/ethereum/web3.py/pull/3705/commits/30f76faa5a0c99c53c96178f48bbd44c0b0f88ff): Found an issue with `PersistentConnectionProvider` batching requests where we hadn't yet separated the send and receive functions to deterministically match responses with their requests by `id`. Fixed that in this PR as well.[](https://github.com/ethereum/web3.py/pull/3705/commits/30f76faa5a0c99c53c96178f48bbd44c0b0f88ff)
- [bugfix](https://github.com/ethereum/web3.py/pull/3705/commits/9cf645dd5cf341810cc8af6b1ec0e8c8dd12ad62): threading id means nothing for async, use loop id
- [test fix](https://github.com/ethereum/web3.py/pull/3705/commits/1542d9df4ac5cf2f0978c05f4d203558705a4245): gather sessions deterministically to check order

---
🎉 🎉  Bonus 🎉 🎉 

I was able to switch our core tests to run via `pytest-xdist` insanely fast by sorting any `set` parametrized params or using `tuple`s / `list`s instead, where appropriate. I am kind of shocked we hadn't gone this route before but this is a pretty significant update.

### Todo:

- [x] Clean up commit history
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)

#### Cute Animal Picture

<img width="968" alt="Screenshot 2025-05-20 at 18 59 53" src="https://github.com/user-attachments/assets/73fadf20-7731-4391-9ca9-1489a6b868ab" />

fselmo committed:
Use ``ContextVar`` for batching context:

- Give some thread safety to batch request context by using
  ``ContextVar`` instead of setting a global state. This actually
  simplifies the logic quite a bit since we don't need to manage
  the batch context state at all.

- Removes ``_is_batching`` from provider classes.

fselmo committed:
Fix test expectations for batching after ``ContextVar`` refactor

fselmo committed:
make core tests able to be run with pytest-xdist:

- Remove sets from parametrization since they can be in random order.
  Either use sorted or change to tuples / lists.

- run with ``-n auto`` and set ``--maxprocesses=15`` for core test runs.

fselmo committed:
Separate batch send from batch receive; cache request funcs:

- Separate the batch send and batch receive functions for persistent
  connection providers in order to deterministically cache the request
  information and be able to retrieve it without any request id guesswork.

- Remove the idea of a batching request counter / id :tada:

- Correct typing expectations for ``BatchRequestInformation``

fselmo committed:
turn off unsupported provider batching tests:

- turn off concurrent batch requests test for LegacyWebsocketProvide
- turn off batching test for EthereumTesterProvider

fselmo committed:
account for loading times in some tests

fselmo committed:
newsfragments for #3705

fselmo committed:
bugfix: threading id means nothing for async, use loop id

- Asynchronous code making use of a threading ``id`` for the cache key
  doesn't make as much sense as using a loop ``id``. Running the core
  tests with ``pytest-xdist`` runners actually seems to have identified
  a bug in this code that was previously well hidden. Replace the threading
  id with loop id when generating unique cache keys in the http session
  manager.

fselmo committed:
test fix: gather sessions deterministically to check order

pacrob commented on tox.ini:
why adding all the space around `=`s?

fselmo replied:
oh weird, must've been my local version of `black`? I'll put this back. We should look to see if we need to update the version of `black` in `make lint` / precommit at some point.

fselmo replied:
wait this is `.ini` so not `black`... something must've updated in my PyCharm formatter ... will fix :/

fselmo replied:
yep this was my fault, I turned something on recently in PyCharm that I didn't think was this but it was. Fixed 👍🏼

fselmo commented on tests/core/providers/test_http_provider.py:
```diff
@@ -127,26 +127,9 @@ def test_http_empty_batch_response(mock_post):
-    "web3._utils.http_session_manager.HTTPSessionManager.make_post_request",
-    new_callable=Mock,
-)
-def test_sync_provider_is_batching_when_make_batch_request(mock_post):
```
~~@pacrob I'm going to put all these back. The original implementation for thread safety didn't lend itself well to this test anymore but I put back `_is_batching` as the `ContextVar` holder / flag so I will put these back now!~~

fselmo replied:
Oh nevermind. So this is actually not required for `provider.make_batch_request()` because the payload that is passed into that is the broken up request information list (list of `("eth_getBlockByNumber", ("latest", False))`), not requests via the API (`w3.eth.get_block("latest")`).

In the recent PR where [we added this](https://github.com/ethereum/web3.py/pull/3663), I was wrong to think we need the batching state while making these requests. We really only need it when we're suppressing the call to the method while calling via the API `w3.eth.get_block()`, etc...

This method should work just fine and your test for `_raise_stray_errors_from_cache` from that PR is all we need I believe.

kclowes commented on tests/integration/go_ethereum/test_goethereum_legacy_ws.py:
```diff
@@ -56,7 +56,10 @@ def w3(start_geth_process_and_yield_port):
+    def test_batch_requests_concurrently_with_regular_requests(
+        self, w3: "Web3"
+    ) -> None:
+        pytest.skip("LegacyWebSocketProvider does not support concurrent requests")
```
nit: maybe use the decorator here to match `test_ethereum_tester` below, and so that it skips before running the test?

fselmo committed:
Clean up unnecessary persistent connection logic for batch:

- Now that we separate send and receive for batching, we no longer have
  to create a workaround for response formatters to contract calls. We
  pass all response formatters together as we do with regular contract
  calls, to be applited to the response as it comes back.

- from comment on PR #3705: move skip marker to decorator

kclowes reviewed (approved):
LGTM! I left one nit, feel free to take or leave!

reedsa reviewed (approved):
lgtm
